### PR TITLE
Dynamic timeline

### LINKED ISSUE #2008: Is there a way to dynamically turn off/on the timeline without killing application ?
Can we turn off the timeline profiler by setting env config. Currently, if I have to profile, I turn the profiler on it writes huge amount of logs and then I have to kill my application. 
Rather than killing application, I would like to dynamically turn it off, by setting a flag or config in a file. 

**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
2. Framework version:
3. Horovod version:
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Please ask your question here.

### PR DESCRIPTION
**Pull request by Vikas-kum**
This PR 
1)introduces a new Python API for starting and stopping the timeline:

```
hvd.start_timeline('/path/to/timeline.json', mark_cycles=True)
...
hvd.stop_timeline()
```
2) Makes sure that timeline json file is valid json. 
3) Including starttime as metadata in timeline.

TODO
 1/ thread detach() re-introduce. 

Fixes #2008.
This is followup from this PR: https://github.com/horovod/horovod/pull/2129 and addresses concurrent access. 
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs?
- [x] Did you write any tests to validate this change?  
- [x] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

Fixes # (issue).

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

**Comment by tgaddair** [comment]
Nice work, @Vikas-kum!  Thanks for picking this up.  I'll take a look today or tomorrow.

File: test/test_torch.py [hunk 1]
```diff
@@ -2133,22 +2134,26 @@ def test_elastic_state(self):
...
-
-            # Perform a simple allreduce operation
+            print("Calling all reduce\n")
             hvd.allreduce(torch.tensor([1, 2, 3], dtype=torch.float32), name='test_allreduce')
-
-            # Wait for timeline to drain before reading it
+            print("Calling stop timeline\n")
+            # stop timeline will immediately stop events to be registered in timeline. We are providing some time
+            # before calling stop so that events can be registered in timeline file.
+            time.sleep(0.2)
```
[the remark this answers is not in the document]
Comment (Vikas-kum) [reply test/test_torch.py:2160]: When we call Start_timeline, timeline is not started immediately and it is started in next cycle of background thread loop(to avoid races and big locks etc.). I put sleep as a caution to make sure that tests remains deterministic.

File: horovod/common/timeline.cc [hunk 2]
```diff
@@ -24,26 +24,99 @@
...
     healthy_ = true;
     active_ = true;
-
-    // Spawn writer thread.
-    writer_thread_ = std::thread(&TimelineWriter::WriterLoop, this);
   } else {
-    LOG(ERROR) << "Error opening the Horovod Timeline file " << file_name
+    LOG(ERROR) << "Error opening the Horovod Timeline file " << filename
                << ", will not write a timeline.";
+    healthy_ = true;
```
[the remark this answers is not in the document]
Comment (Vikas-kum) [reply horovod/common/timeline.cc:94]: Yes active_ is main thing controlling timeline acceptance of event. healthy_ = true is indication that a thread is always active, goes false only in Shutdown method. 
active_ is real control which decides whether timeline should accept event and then write event to file.

File: horovod/common/timeline.cc [hunk 3]
```diff
@@ -76,7 +149,15 @@ void TimelineWriter::EnqueueWriteMarker(const std::string& name,
...
 void TimelineWriter::DoWriteEvent(const TimelineRecord& r) {
   assert(r.type == TimelineRecordType::EVENT);
-
+  if(is_new_file_){
+    file_ << "[\n";
+    is_new_file_ = false;
+  } else {
+    // last event closed the json ']' , need to seek to one position back and write ',' to continue
+    long pos = file_.tellp();
+    file_.seekp (pos-1);
```
[the remark this answers is not in the document]
Comment (Vikas-kum) [reply horovod/common/timeline.cc:179]: I dont think it should cause any perf penalty, given that we are already going to write bytes from next position. 
Would you be able to help me with quickly running a small job and verifying on this PR as I will ahve to go through setups, which i dont currently have. :) I know I am asking for a favor but will appreciate if you can help.

**Comment by Vikas-kum** [comment]
Thanks @tgaddair for review. Will update PR soon addressing comments.

File: horovod/common/operations.cc [hunk 4]
```diff
@@ -415,18 +415,26 @@ void BackgroundThreadLoop(HorovodGlobalState& state) {
 
   // Open the timeline file on coordinator.
   auto horovod_timeline = std::getenv(HOROVOD_TIMELINE);
-  if (is_coordinator && horovod_timeline != nullptr) {
+  if (horovod_timeline == nullptr) {
```
[the remark this answers is not in the document]
Comment (Vikas-kum) [reply horovod/common/operations.cc]: Done

File: test/test_torch.py [hunk 5]
```diff
@@ -2133,22 +2134,26 @@ def test_elastic_state(self):
...
 
         with temppath() as fname1:
             hvd.start_timeline(fname1, mark_cycles=True)
-
-            # Perform a simple allreduce operation
+            print("Calling all reduce\n")
             hvd.allreduce(torch.tensor([1, 2, 3], dtype=torch.float32), name='test_allreduce')
-
-            # Wait for timeline to drain before reading it
+            print("Calling stop timeline\n")
```
[the remark this answers is not in the document]
Comment (Vikas-kum) [reply test/test_torch.py]: Done

File: test/test_torch.py [hunk 6]
```diff
@@ -2133,22 +2134,26 @@ def test_elastic_state(self):
...
             if hvd.rank() == 0:
                 with open(fname, 'r') as timeline_file:
                     timeline_text = timeline_file.read()
+                    print("Timeline text is : ")
+                    print(timeline_text + "\n")
                     assert 'allreduce.test_allreduce' in timeline_text, timeline_text
                     assert 'NEGOTIATE_ALLREDUCE' in timeline_text, timeline_text
                     assert 'ALLREDUCE' in timeline_text, timeline_text
-                    assert 'CYCLE_START' in timeline_text, timeline_text
+                    if check_cycle is True:
```
[the remark this answers is not in the document]
Comment (Vikas-kum) [reply test/test_torch.py]: Done

File: horovod/common/timeline.cc [hunk 7]
```diff
@@ -24,26 +24,99 @@
...
+    cur_filename_ = filename;
+    new_pending_filename_ = cur_filename_;
+    // empty filename is special which tells that init the timeline but don't activate it.
+    healthy_ = true;
+    active_ = false;
+    LOG(INFO) << "Inited TimelineWriter but active_ is false, since filename passed is empty string";
+    return;
+  }
+  // if newfile is same as existing file, it should be no-op and we can continue writing in existing file.
+  if(cur_filename_ == filename) {
```
[the remark this answers is not in the document]
Comment (Vikas-kum) [reply horovod/common/timeline.cc]: Done. I only formatted timeline.cc and timeline.h . As there are so many other changes getting introduced if I format other files.

**Comment by Vikas-kum** [comment]
@tgaddair Gentle reminder. Would you be take another round at this.

File: test/test_torch.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tgaddair) [reply test/test_torch.py:2156]: I think we should make these calls blocking.  Should be simple enough to do by capturing the change of the pending flag.  Does that sounds reasonable to you?

**Review by tgaddair — COMMENTED** [review]
I really like the changes to Timeline to produce correct JSON.

Rather than make the Timeline object thread safe, I think it's simpler to just place a mutex around the start and stop timeline functions.  This also allows us to make the functions blocking, which seems referable to me.  I just added this to the other PR here: https://github.com/horovod/horovod/pull/2129/commits/3b1a85a786acd51b064a0e352a7863bb495fdb80

If that passes, and those changes sound reasonable to you, then I would propose making this PR just about the changes to the Timeline itself (fixing JSON, allowing timeline files to be reused, etc.).

What do you think?

**Comment by Vikas-kum** [comment]
>>
The one thing I would still like to see is to make these calls blocking so the user doesn't need to sleep after each call (otherwise, they could miss allreduce information on the timeline when starting, or the end of the timeline if they stop and then exit the program). Does that sound reasonable to you?

I took this approach because it simplifies the implementation quite a bit and also makes sure that locks are held for much smaller and confined scope. 
We can make stop timeline blocking, but even then we would need sleep in test code to ensure that events are written. 
Say you start_timeline

Start takes into effect only when BGThread goes to next iteration. To make sure in test checks for cycle markers and events getting captured we would need to wait for at least couple of cycles. 

hvd.start_Timeline()--> (sleep here in test code makes sure that timeline is started before next line test code)starts takes in to effect --> events are written and cycles are written

Somehow, I don't like the idea of blocking call because it blocks my training scripts completely. I do see a side effect of having non-blocking calls that for one cycle events wont be written. But at this point, I expect user would be calling start_timeline to stop_timeline for sufficient gap which gives them a good picture of what is happening. 

For this reason, I would like to have this non-blocking. Please let me know if you have strong feeling about blocking call and I can make stop timeline blocking.

File: test/test_torch.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Vikas-kum) [reply test/test_torch.py:2156]: Done! Made both start_timeline and stop_timeline calls blocking.

**Comment by tgaddair** [comment]
Thanks for making these changes, @Vikas-kum!  

Last thing I would ask: can run `clang-format` from the root of the project to make sure code style aligns with the rest of the code base?

If tests pass after that, we should be good to land :)

**Comment by Vikas-kum** [comment]
> Thanks for making these changes, @Vikas-kum!
> 
> Last thing I would ask: can run `clang-format` from the root of the project to make sure code style aligns with the rest of the code base?
> 
> If tests pass after that, we should be good to land :)

thanks @tgaddair , I ran clang-format for only timeline.h timeline.cc 
I see running clang-format for other files introduces a lot of formatting changes. So, I would suggest that we should have different PR just with clang-format run. WDYT?

**Review by tgaddair — APPROVED** [review]
Sounds good, will take a look at the follow-up PR!

**Comment by Vikas-kum** [comment]
> Sounds good, will take a look at the follow-up PR!

Thanks. Created this PR: https://github.com/horovod/horovod/pull/2281/files , 
Ran only for horovod common to keep the PR manageable for review.

**Comment by mengxr** [comment]
@Vikas-kum We found an issue introduced by this commit. Our distributed filesystem doesn't support random writes (including re-open and append). It used to work fine with horovod timeline because it only create, writes the data sequentially, and then close. I wonder if we can keep the old behavior if users do not use the dynamic timeline feature. Any pointers on what we should change?
